Phone sign in

bdlukaa opened the pull request:
## What kind of change does this PR introduce?

Feature.

Fixes https://github.com/supabase/supabase-dart/issues/38

## What is the new behavior?

Phone authentication is added. Added the following methods:

- `signUpWithPhone`
- `signInWithPhone`
- `sendMobileOTP`
- `verifyMobileOTP`

bdlukaa commented on lib/src/gotrue_api.dart:
```diff
@@ -75,6 +75,62 @@ class GoTrueApi {
+  /// `phone` is the user's phone number WITH international prefix
+  ///
+  /// `password` is the password of the user
+  Future<GotrueSessionResponse> signUpWithPhone(
```
https://github.com/supabase/gotrue-js/blob/797ea09b6188291861b5bed4a488cc74b6f95aaf/src/GoTrueApi.ts#L94

bdlukaa commented on lib/src/gotrue_api.dart:
```diff
@@ -75,6 +75,62 @@ class GoTrueApi {
+  /// `phone` is the user's phone number WITH international prefix
+  ///
+  /// `password` is the password of the user
+  Future<GotrueSessionResponse> signInWithPhone(
```
https://github.com/supabase/gotrue-js/blob/797ea09b6188291861b5bed4a488cc74b6f95aaf/src/GoTrueApi.ts#L114

bdlukaa commented on lib/src/gotrue_api.dart:
```diff
@@ -100,6 +156,55 @@ class GoTrueApi {
+  /// Sends a mobile OTP via SMS. Will register the account if it doesn't already exist
+  ///
+  /// `phone` is the user's phone number WITH international prefix
+  Future<GotrueJsonResponse> sendMobileOTP(String phone) async {
```
https://github.com/supabase/gotrue-js/blob/797ea09b6188291861b5bed4a488cc74b6f95aaf/src/GoTrueApi.ts#L158

bdlukaa commented on lib/src/gotrue_api.dart:
```diff
@@ -100,6 +156,55 @@ class GoTrueApi {
+  /// `phone` is the user's phone number WITH international prefix
+  ///
+  /// `token` is the token that user was sent to their mobile phone
+  Future<GotrueSessionResponse> verifyMobileOTP(String phone, String token,
```
https://github.com/supabase/gotrue-js/blob/797ea09b6188291861b5bed4a488cc74b6f95aaf/src/GoTrueApi.ts#L174

bdlukaa commented:
No idea why tests are failing

dshukertjr commented:
@bdlukaa Amazing work 🎉

Test failing seems like a separate issue, so I don't think you need to worry about it for now. 

You probably already on it, but could we also add these?

Could we also add `phone` property to `User`? [gotrue-js implementation](https://github.com/supabase/gotrue-js/pull/106/files#diff-b8b77f5be18cf56dca425b3a5b8e9d2e754dd37fe0e3612b95cd4e9bccda08a5R49)

And add phone option to `signIn()` [gotrue-js implementation](https://github.com/supabase/gotrue-js/pull/106/files#diff-3522461172efd6058d6b8da62fc2d30d8b524d2b64894ea2c67218c52f7fdff5L152-L182)

And add phone option to `signUp()` [gotrue-js implementation](https://github.com/supabase/gotrue-js/pull/106/files#diff-3522461172efd6058d6b8da62fc2d30d8b524d2b64894ea2c67218c52f7fdff5L97-L114)
Except this would be a breaking change for us... Should we create a separate `signUpWithPhone()` method?

Add `verifyOTP()` method to `GoTrueClient` [gotrue-js implementation](https://github.com/supabase/gotrue-js/pull/106/files#diff-3522461172efd6058d6b8da62fc2d30d8b524d2b64894ea2c67218c52f7fdff5R222-R242)

bdlukaa commented:
@dshukertjr Done!

Added the following methods to `GoTrueClient`:

- `signUpWithPhone`
- `verifyOTP`
- `_handlePhoneSignIn`

And updated `signIn` with support for Phone auth.

Added the following properties to `User`:

- `phone`
- `emailConfirmedAt`
- `phoneConfirmedAt`

and deprecated `confirmedAt`

bdlukaa commented:
I have not tested the PR in any project (yet). If anyone could test it to see if it actually works, it would be awesome!

phamhieu commented:
Oh wow, that's fast :rocket: :rocket: :rocket: 

> No idea why tests are failing

We enforce line length rule check `Run dart format lib test -l 80 --set-exit-if-changed` It seems one of the file needs format

dshukertjr commented:
@bdlukaa 
It seems like there are some code that got deprecated in this PR being used in the code base. If we could sort these out, we should be able to merge this!

```
info - lib/src/gotrue_client.dart:65:30 - 'confirmedAt' is deprecated and shouldn't be used. Use emailConfirmedAt. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use_from_same_package
info - lib/src/gotrue_client.dart:315:30 - 'confirmedAt' is deprecated and shouldn't be used. Use emailConfirmedAt. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use_from_same_package
info - lib/src/user.dart:57:25 - 'confirmedAt' is deprecated and shouldn't be used. Use emailConfirmedAt. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use_from_same_package
```

bdlukaa commented:
Could you test the PR, please? I hadn't have the time to test it.

Yes, we should migrate to the new options

Saravia83c commented:
#26

Saravia83c commented:
Duplicate of #

dshukertjr commented:
@bdlukaa 
Sure, I will test it during this weekend!

dshukertjr commented:
Signup with phone number, verifying  OTP, signing in with phone number and password, signing in with phone number and OTP all worked 🎉 Amazing work @bdlukaa !

dshukertjr commented:
Yeah, the tests are failing for weird, seemingly unrelated reasons again... 
@phamhieu I think there was a similar issue with it before as well, but do you know what the cause was in that case?
https://github.com/supabase/gotrue-dart/commit/fe12bce27d9b79574e3413c515b0039ba2423b38

phamhieu commented:
I update infra test with latest postgres image from https://hub.docker.com/r/supabase/postgres/tags?page=1&ordering=last_updated

Can you merge and check again?

dshukertjr commented:
Thank you so much @phamhieu 🎉
